### PR TITLE
Add istio.io test support for mixed content scripts

### PR DESCRIPTION
**Pull request by nmittler**
This PR adds support for reading scripts that contain
a mix of bash scripts and snippets.

The new ScriptParser step creates a single Command
from the input, but can generate a number of Snippets.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

### CODE REVIEW

File: pkg/test/istioio/script_parser.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+	snippetEndToken   = "$endsnippet"
+
+	syntaxKey   = "syntax"
+	excludeKey  = "exclude"
+	outputisKey = "outputis"
+)
+
+var _ Step = ScriptParser{}
+
+// ScriptParser is a test Step that parses an input script which may contain a mix of raw bash commands
```
Comment (geeknoid) [inline pkg/test/istioio/script_parser.go]: Er, I'm a bit confused by the need for this.

The original design with snippets was that this combination was directly supported by the web site infra without any special magic, just by using comments:

```
# $snippet SNIP1
cmd1
# $endsnippet

# $snippet SNIP2
cmd3
# $endsnippet
```

So you can run this script as-is with no special parser. On the ingestion side in istio.io, the regex can pluck out the snippets in the usual way.

Is the ScriptParser giving more functionality than that?

Comment (nmittler) [reply pkg/test/istioio/script_parser.go]: Ah that's a good point ... I had forgotten about the support for `# $snippet...`. So the idea would be that the entire script would be executable and we could effectively copy the content of the entire file over to the `*.snippets.txt` file.

Let me think about it some more.

**Comment by brian-avery** [comment]
This will need readme updates. 

I love the idea of this because it means that those who write the docs don't have to worry about the Go code as much when they write their tests (not all docs writers are programmers). I wonder if we should leave the command option for now though. We have users writing istio.io tests and should standardize on what the interface looks like to them.

File: tests/integration/istioio/security/scripts/authz_http.txt [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+set -u
+set -o pipefail
+
+REPEAT=${REPEAT:-100}
+THRESHOLD=${THRESHOLD:-20}
+
+# verify calls curl to send requests to productpage via ingressgateway.
+# - The 1st argument is the expected http response code
+# - The remaining arguments are the expected text in the http response
+# Return 0 if both the code and text is found in the response for continuously $THERESHOLD times,
```
Comment (brian-avery) [inline tests/integration/istioio/security/scripts/authz_http.txt]: THRESHOLD

**Comment by nmittler** [comment]
@brian-avery 

> I wonder if we should leave the command option for now though. We have users writing istio.io tests and should standardize on what the interface looks like to them.

I think I'd prefer just removing it wholesale now before we get too many tests using it.

**Comment by nmittler** [comment]
@brian-avery PTAL at the readme updates.

File: pkg/test/istioio/script.go [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+func filterCommandLine(line string) string {
+	// Remove surrounding @'s for github links.
+	line = githubLinkMatch.ReplaceAllStringFunc(line, func(input string) string {
+		return input[1 : len(input)-1]
+	})
+
+	// Remove any command line prefixes.
+	line = strings.TrimPrefix(line, commandLinePrefix)
+
+	// Remove any surrounding whitespace.
```
Comment (brian-avery) [inline pkg/test/istioio/script.go]: Are you missing something here?

File: pkg/test/istioio/verifier.go [hunk 4]
```diff
@@ -20,19 +20,11 @@ import (
 	"unicode"
 )
 
-// Verifier is a function used to verify output and any errors returned.
-type Verifier func(ctx Context, name, output string, err error)
+// verifier is a function used to verify output and any errors returned.
+type verifier func(ctx Context, name, output string, err error)
```
[the remark this answers is not in the document]
Comment (nmittler) [reply pkg/test/istioio/verifier.go]: we can do that when we have a need.

File: pkg/test/istioio/script.go (hunk 3 above)
Comment (nmittler) [reply pkg/test/istioio/script.go]: outdated comment.

**Comment by nmittler** [comment]
@brian-avery PTAL

File: pkg/test/istioio/verifier.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (brian-avery) [reply pkg/test/istioio/verifier.go]: Sounds good.

File: tests/integration/istioio/README.md [hunk 5]
```diff
@@ -146,35 +140,126 @@ istioio.IfMinikube{
 }
 ```
 
-## Run Shell Commands
+## Running Shell Commands
 
-You can create a test step that will run a shell command with `istioio.Command`:
+You can create a test step that will run a bash script and automatically generate
```
Comment (brian-avery) [inline tests/integration/istioio/README.md]: Do we still have zsh and other shells discussed in the Istio? I know we did at some point.

Comment (brian-avery) [reply tests/integration/istioio/README.md]: Nit, just thinking about how we're specifying bash.

Comment (brian-avery) [reply tests/integration/istioio/README.md]: Looks like Kiali and auto completion talk about it.

**Comment by brian-avery** [comment]
@nmittler These changes need to be done to the mirror test as well.

File: tests/integration/istioio/security/mtls_migration_test.go [hunk 6]
```diff
@@ -30,147 +30,44 @@ func TestMutualTLSMigration(t *testing.T) {
 	framework.
 		NewTest(t).
 		Run(istioio.NewBuilder("tasks__security__mututal_tls_migration").
-			Add(istioio.Command{
+			Add(istioio.Script{
```
Comment (brian-avery) [inline tests/integration/istioio/security/mtls_migration_test.go]: These changes need to be done to mirror_test as well.

**Pull request merged**